src/state.rs

```diff
@@ -134,41 +176,43 @@ impl<'a> StateMachine<'a> {
     }
 
     /// Removes the current state on the stack and inserts a different one.
-    fn switch(&mut self, state: Box<State>, engine: &mut Engine) {
+    fn switch(&mut self, mut state: Box<State>, engine: &mut Engine) {
         if self.running {
-            if let Some(mut state) = self.state_stack.pop() {
-                state.on_stop(engine);
+            if let Some((ref mut state, _, ref mut scene)) = self.state_stack.pop() {
+                state.on_stop(engine, scene);
+                engine.world.delete_entities(&scene.vec);
             }
 
-            self.state_stack.push(state);
-            let state = self.state_stack.last_mut().unwrap();
-            state.on_start(engine);
+            let mut scene = Scene::new();
+            let dispatcher = state.on_start(engine, &mut scene).map(|db| db.with_pool(engine.pool.clone()).build());
+            self.state_stack.push((state, dispatcher, scene));
         }
     }
 
     /// Pauses the active state and pushes a new state onto the state stack.
-    fn push(&mut self, state: Box<State>, engine: &mut Engine) {
+    fn push(&mut self, mut state: Box<State>, engine: &mut Engine) {
         if self.running {
-            if let Some(state) = self.state_stack.last_mut() {
-                state.on_pause(engine);
+            if let Some(&mut (ref mut state, _, ref mut scene)) = self.state_stack.last_mut() {
+                state.on_pause(engine, scene);
             }
 
-            self.state_stack.push(state);
-            let state = self.state_stack.last_mut().unwrap();
-            state.on_start(engine);
+            let mut scene = Scene::new();
+            let dispatcher = state.on_start(engine, &mut scene).map(|db| db.with_pool(engine.pool.clone()).build());
```

Comment:
The same `State` will be pushed and popped again and again during the game.
Building dispatchers each time may be slow. Can't we cache 'em in some way?

Reply:
We could, but that introduces its own problems because any state the systems may contain will be preserved between instances of the state.  Why would building a dispatcher be slow?  It doesn't depend on external data or operate on large data sets.